packages/api/src/WithServerEvents.tsx

```diff
@@ -0,0 +1,241 @@
+/* tslint:disable */
+import * as React from 'react';
+import { IHeaders } from '@syndesis/api';
+import { resolve as _resolve } from 'url';
+
+export interface IChangeEvent {
+  action: string;
+  kind: string;
+  id: string;
+}
+
+export interface IMessageEvent {
+  id: string;
+  data: string;
+  event: string;
+}
+
+export interface IWithEventsProps {
+  apiUri: string;
+  headers: IHeaders;
+  children(props: IWithEventsState): any;
+}
+
+export interface IWithEventsState {
+  registerChangeListener: (
+    key: string,
+    listener: (event: IChangeEvent) => void
+  ) => void;
+  registerMessageListener: (
+    key: string,
+    listener: (event: IMessageEvent) => void
+  ) => void;
+  unregisterChangeListener: (key: string) => void;
+  unregisterMessageListener: (key: string) => void;
+}
+
+const RECONNECT_TIME = 5000;
+
+export class WithServerEvents extends React.Component<
+  IWithEventsProps,
+  IWithEventsState
+> {
+  private starting = false;
+  private started = false;
+  private retries = 0;
+  private preferredProtocol: string = '';
+  private eventSource: EventSource | undefined;
+  private webSocket: WebSocket | undefined;
+  private apiEndpoint: string | undefined;
+  private headers: IHeaders | undefined;
+  private changeListeners = {};
+  private messageListeners = {};
+
+  public constructor(props: IWithEventsProps) {
+    super(props);
+    this.state = {
+      registerChangeListener: this.registerChangeListener.bind(this),
+      registerMessageListener: this.registerChangeListener.bind(this),
+      unregisterChangeListener: this.unregisterChangeListener.bind(this),
+      unregisterMessageListener: this.unregisterMessageListener.bind(this),
+    };
+  }
+
+  public registerChangeListener(
+    key: string,
+    listener: (event: IChangeEvent) => void
+  ) {
+    this.changeListeners[key] = listener;
+  }
+
+  public registerMessageListener(
+    key: string,
+    listener: (event: IMessageEvent) => void
+  ) {
+    this.messageListeners[key] = listener;
+  }
+
+  public unregisterChangeListener(key: string) {
+    delete this.changeListeners[key];
+  }
+
+  public unregisterMessageListener(key: string) {
+    delete this.messageListeners[key];
+  }
+
+  public async componentDidMount() {
+    this.apiEndpoint = this.props.apiUri;
+    this.headers = this.props.headers;
```

Comment:
Here you are just duplicating the props without doing any change. Props should be considered a single source of truth, storing their value in some internal state is a common pattern for bugs. Just refer to the `this.props.propName` value when you need it.